Color picker

Banana090 opened the pull request:
Kinda lost track of all the updates this PR contains, so... It's just huge.
For in-game changes it changes HUD settings tab and adds Color Picker UI feature with preview, editboxes on RMB and so on...

https://cdn.discordapp.com/attachments/295908390956433410/787843848810135582/unknown.png

(Github does not want to upload my image here...)

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- commented:
I think you forgot to `git add` some files, doesn't build for me:
```
In file included from src/game/client/components/console.cpp:30:
src/game/client/components/menus.h:82:107: error: ‘SUIAnimator’ has not been declared
   82 |  int DoButton_MenuTab(const void *pID, const char *pText, int Checked, const CUIRect *pRect, int Corners, SUIAnimator *pAnimator = NULL, const ColorRGBA *pDefaultColor = NULL, const ColorRGBA *pActiveColor = NULL, const ColorRGBA *pHoverColor = NULL, float EdgeRounding = 10, int AlignVertically = 1);
      |                                                                                                           ^~~~~~~~~~~
```

Banana090 commented:
Yes, something probably went wring with merge, will check...

Banana090 commented:
@def- fixed

def- commented:
Right click to edit the hex value feels a bit unintuitive, I think left click would be better.

Escape also closes the entire menu and brings you back to main menu, not sure i that should be fixed, I expected it to close the color selection.

Strings like "Friend" and "Spammer" in the preview window could be localized.

There is also a color selection in Settings -> DDNet for the backgrounds, could they use the color picker too?

Haven't read the code yet.

Banana090 commented:
@def- yes, everything can use this easily until you're using alpha or clamping (tee skin color selector). These features will be added in the future.
I've implemented left click en HEX first, but then it ended up with some bugs that... I suppose... Were fixed already, but I thought left clicking only works for HEX is kinda weird. We can try this anyway.

For the ESC I think it's better to stay with current behaviour as changing It could be unintuitive, but I can also implement this if you want. (For now you can close color selector vía right click anywhere or left click outside of selector. Or by switching tab)

Banana090 commented:
1. Settings->DDNet backgrounds color now being chosen via Color Picker
2. Fixed bug when you exit menu via ESC, then open it again and color picker was still active
3. Fixed bug when you switch tab using Mouse3 button and color picker was still active

Banana090 commented:
Now I would like to move on to the next features. Code for color picker could be improved of course, but for me it doesn't worth the effort. Color picker seems to work fine for me, the only thing is maybe Left-Clickable HEX value, but I've already tested it and I did not like it. Also it produces some bugs which must be fixed, so I don't want to do this xD

I don't want to stuck here adding very small changes for a long time since it will burn out my motivation to make other updates, so...

Feel free to **@at** me if you find any critical issues, I will fix them. Or if you need any help / explanations of code. I'll go make some other cool updates

P.S. For conflict resolving use versions from `pr_animated_tabs`

def- commented:
I don't like the way it looks in DDNet Settings menu:
![screenshot-20201216@091034](https://user-images.githubusercontent.com/2335377/102321901-96997280-3f7e-11eb-8bab-6f98c528990c.png)
Not sure how best to change it though.

def- commented:
> I don't want to stuck here adding very small changes for a long time since it will burn out my motivation to make other updates, so...

So we should instead burn out our motivation fixing your PRs?

Learath2 commented:
> I don't like the way it looks in DDNet Settings menu:
> 
> Not sure how best to change it though.

Hm, maybe just left aligning might help but I think it's a missing headline that's throwing us off. Maybe "UI Colors"?

def- commented:
bors r+
Better than not having it at least.

def- commented:
bors r+

Banana090 commented:
Some polishing for you guys
![Screenshot_2](https://user-images.githubusercontent.com/39315809/102658656-0713df80-4189-11eb-8b57-43751684260c.png)
😘

def- commented:
bors r+
Thanks

Jupeyy commented on src/game/client/components/menus_settings.cpp:
```diff
@@ -1587,430 +1642,285 @@ ColorHSLA CMenus::RenderHSLScrollbars(CUIRect *pRect, unsigned int *pColor, bool
+	TextRender()->MoveCursor(&Cursor, RealMsgPaddingTee, 0);
+	TextRender()->TextColor(DefaultNameColor);
+	if(g_Config.m_ClShowIDs)
+		TextRender()->TextEx(&Cursor, " 7: Evgesha: ", -1);
```
And here

Jupeyy commented on src/game/client/components/menus_settings.cpp:
```diff
@@ -1587,430 +1642,285 @@ ColorHSLA CMenus::RenderHSLScrollbars(CUIRect *pRect, unsigned int *pColor, bool
 
-		ColorHSLA LOColor = RenderHSLScrollbars(&Laser, &g_Config.m_ClLaserOutlineColor);
+		if(g_Config.m_ClShowIDs)
+			str_copy(LineBuilder, " 7: Evgesha: Hey, how are you ", sizeof(LineBuilder));
```
Could we please use a neutral name and maybe a skin that is shipped with the client

Banana090 replied:
Of course not)

Jupeyy replied:
I dont think the client should trigger skindownloads if not needed.
And the problem with a random name is, that it might sound similar to a word in a different language

Jupeyy commented on src/game/client/components/menus_settings.cpp:
```diff
@@ -1587,430 +1642,285 @@ ColorHSLA CMenus::RenderHSLScrollbars(CUIRect *pRect, unsigned int *pColor, bool
 
-		Graphics()->QuadsEnd();
+		RenderInfo[i++].m_OriginalRenderSkin = GameClient()->m_pSkins->Get(DefaultInd)->m_OriginalSkin;
+		RenderInfo[i++].m_OriginalRenderSkin = (ind = GameClient()->m_pSkins->Find("planet_uranus")) != -1 ? GameClient()->m_pSkins->Get(ind)->m_OriginalSkin : RenderInfo[0].m_OriginalRenderSkin;
```
the skin
